internal/locate/replica_selector.go

```diff
@@ -0,0 +1,486 @@
+	return nil
+}
+
+func (s ReplicaSelectMixedStrategy) isCandidate(r *replica, isLeader bool, epochStale bool, liveness livenessState) bool {
+	if epochStale || liveness == unreachable {
+		// the replica is not available, skip it.
+		return false
+	}
+	maxAttempt := 1
+	if r.dataIsNotReady {
+		// If the replica is failed by data not ready with stale read, we can retry it with replica-read.
+		maxAttempt = 2
+	}
+	if r.isExhausted(maxAttempt, 0) {
+		// attempts is exhausted, skip it.
+		return false
+	}
+	if s.leaderOnly && !isLeader {
+		return false
+	}
+	if s.busyThreshold > 0 && (r.store.EstimatedWaitTime() > s.busyThreshold || r.serverIsBusy || isLeader) {
+		return false
+	}
+	return true
+}
+
+const (
+	// define the score of priority.
+	scoreOfLabelMatch   = 3
+	scoreOfPreferLeader = 2
+	scoreOfNormalPeer   = 1
+	scoreOfNotSlow      = 1
+)
+
+func (s *ReplicaSelectMixedStrategy) calculateScore(r *replica, isLeader bool) int {
+	score := 0
+	if isLeader {
+		if s.preferLeader {
+			score += scoreOfPreferLeader
+		} else if s.tryLeader {
+			if len(s.labels) > 0 {
+				// when has match labels, prefer leader than not-matched peers.
+				score += scoreOfPreferLeader
+			} else {
+				score += scoreOfNormalPeer
+			}
+		}
+	} else {
+		if s.learnerOnly {
+			if r.peer.Role == metapb.PeerRole_Learner {
+				score += scoreOfNormalPeer
+			}
+		} else {
+			score += scoreOfNormalPeer
+		}
+	}
+	if r.store.IsStoreMatch(s.stores) && r.store.IsLabelsMatch(s.labels) {
+		score += scoreOfLabelMatch
+	}
+	if !r.store.isSlow() {
+		score += scoreOfNotSlow
+	}
+	return score
+}
+
+type ReplicaSelectLeaderWithProxyStrategy struct{}
+
+func (s ReplicaSelectLeaderWithProxyStrategy) next(replicas []*replica, region *Region) (leader *replica, proxy *replica) {
+	rs := region.getStore()
+	leaderIdx := rs.workTiKVIdx
+	leader = replicas[leaderIdx]
+	if leader.store.getLivenessState() == reachable {
+		// if leader's store is reachable, no need use proxy.
+		rs.unsetProxyStoreIfNeeded(region)
+		return nil, nil
+	}
+	proxyIdx := rs.proxyTiKVIdx
+	if proxyIdx >= 0 && int(proxyIdx) < len(replicas) && s.isCandidate(replicas[proxyIdx], proxyIdx == leaderIdx) {
+		return leader, replicas[proxyIdx]
+	}
+
+	for i, r := range replicas {
+		if s.isCandidate(r, AccessIndex(i) == leaderIdx) {
+			return leader, r
+		}
+	}
+	return nil, nil
+}
+
+func (s ReplicaSelectLeaderWithProxyStrategy) isCandidate(r *replica, isLeader bool) bool {
+	if isLeader ||
+		r.isExhausted(1, 0) ||
+		r.store.getLivenessState() != reachable ||
+		r.isEpochStale() {
+		// check epoch here, if epoch staled, we can try other replicas. instead of buildRPCContext failed and invalidate region then retry.
+		return false
+	}
+	return true
+}
+
+func (s *replicaSelectorV2) onNotLeader(
+	bo *retry.Backoffer, ctx *RPCContext, notLeader *errorpb.NotLeader,
+) (shouldRetry bool, err error) {
+	if s.target != nil {
+		s.target.notLeader = true
```

Comment:
Yes. If not set `target.notLeader` to true, then will retry this replica, and this behavior is not compatible with before, since it should retry the remaining replica.